docfx_yaml/extension.py

```diff
@@ -1417,29 +1417,94 @@ def find_markdown_pages(app, outdir):
                 'href': mdfile.name.lower(),
             })
 
+def check_existing_references(
+    current_name: str,
+    words: List[str],
+    index: int,
+    word: str,
+    keyword: str,
+    converted_words: List[str],
+    hard_coded_references: Dict[str, str] = None
+):
+    """Given a word, check if it contains a reference keyword we should convert.
+
+    Args:
+        current_name: the name of the current Python object being looked at
+        words: list of words used to check and compare content before and after the given word
+        index: current index within words
+        word: current word being looked at
+        keyword: reference keyword to check against
+        converted_words: list of words containing words that's been processed so far
+        hard_coded_references: Optional list containing a list of hard coded references
+
+    Returns:
+        Empty string if no reference needs to be converted, or a string that contains
+          the converted reference.
+
+    """
+    cross_reference = ""
+    if keyword != current_name and keyword not in current_name and keyword in word:
+        if keyword in hard_coded_references:
+            if "<a" not in words[index-1] and \
+                not (converted_words and f"<a href=\"{hard_coded_references[keyword]}" in converted_words[-1]):
+                cross_reference = f"<a href=\"{hard_coded_references[keyword]}\">{keyword}</a>"
+
+        else:
+            # If it is already processed as cross reference, skip over it.
+            if "<xref" not in words[index-1] and \
+                not (converted_words and f"<xref uid=\"{keyword}" in converted_words[-1]):
+                cross_reference = f"<xref uid=\"{keyword}\">{keyword}</xref>"
+
+    return cross_reference
+
 
-# Finds and replaces occurrences which should be a cross reference in the given
-# content, except for the current name.
 def convert_cross_references(content: str, current_name: str, entry_names: List[str]):
+    """Finds and replaces references that should be a cross reference in given content.
+
+    This should not convert any references that contain its name, i.e. if we're processing
+    `google.cloud.spanner.v1.services`, references to `google.cloud.spanner.v1.services`
+    should not be convereted to references.
+
+    Args:
+        content: given body of content to parse and look for references
+        current_name: the name of the current Python object being looked at
+        entry_names: list of keyword references to look for
+
+    Returns:
+        content that has been modified with proper cross references if found.
+
+    """
     words = content.split(" ")
-    new_words = []
+    converted_words = []
```

Comment:
Why a list instead of a dictionary from word to converted word?

Reply:
I suppose the name is still throwing off. As we traverse the original list of words, we still need to check against the converted list of words and order matters. At the end, I use this to simply convert to a string using a join. Will update the name so it's not misleading.